### PR TITLE
feat: implement maxEB EIP-7251

### PR DESCRIPTION
**Pull request by ensi321**
maxeb impl for devnet0

### CODE REVIEW

File: packages/beacon-node/src/chain/chain.ts [hunk 1]
```diff
@@ -927,6 +927,9 @@ export class BeaconChain implements IBeaconChain {
     metrics.forkChoice.balancesLength.set(forkChoiceMetrics.balancesLength);
     metrics.forkChoice.nodes.set(forkChoiceMetrics.nodes);
     metrics.forkChoice.indices.set(forkChoiceMetrics.indices);
+
+    const headState = this.getHeadState();
+    metrics.headState.unfinalizedPubkeyCacheSize.set(headState.epochCtx.unfinalizedPubkey2index.size);
```
Comment (twoeths) [inline packages/beacon-node/src/chain/chain.ts]: this should be for EIP-6110 instead?

File: packages/beacon-node/src/chain/regen/queued.ts [hunk 2]
```diff
@@ -194,6 +194,54 @@ export class QueuedStateRegenerator implements IStateRegenerator {
     return this.checkpointStateCache.updatePreComputedCheckpoint(rootHex, epoch);
   }
 
+  /**
+   * Remove `validators` from all unfinalized cache's epochCtx.UnfinalizedPubkey2Index,
+   * and add them to epochCtx.pubkey2index and epochCtx.index2pubkey
+   */
+  updateUnfinalizedPubkeys(validators: UnfinalizedPubkeyIndexMap): void {
```
Comment (twoeths) [inline packages/beacon-node/src/chain/regen/queued.ts]: looks like this branch includes EIP-6110 code @ensi321

File: packages/state-transition/src/block/processDeposit.ts [hunk 3]
```diff
@@ -41,68 +49,132 @@ export function processDeposit(fork: ForkSeq, state: CachedBeaconStateAllForks,
...
+    if (fork < ForkSeq.electra) {
+      // increase balance by deposit amount right away pre-electra
+      increaseBalance(state, cachedIndex, amount);
+    } else if (fork >= ForkSeq.electra) {
+      const stateElectra = state as CachedBeaconStateElectra;
+      const pendingBalanceDeposit = ssz.electra.PendingBalanceDeposit.toViewDU({
+        index: cachedIndex,
+        amount: BigInt(amount),
+      });
+      stateElectra.pendingBalanceDeposits.push(pendingBalanceDeposit);
```
Comment (twoeths) [inline packages/state-transition/src/block/processDeposit.ts:83]: so we add new `PendingBalanceDeposit` with `amount` but inside `switchToCompoundingValidator()` we also add another excess `PendingBalanceDeposit` with `balance - MIN_ACTIVATION_BALANCE`

is it correct to add 2 `PendingBalanceDeposit` for a single deposit? it's not an issue of our implementation just a question for the spec

File: packages/state-transition/src/block/processExecutionLayerWithdrawalRequest.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  }
+}
+
+function isValidValidator(
+  validator: phase0.Validator,
+  sourceAddress: Uint8Array,
+  state: CachedBeaconStateElectra
+): boolean {
+  const {withdrawalCredentials} = validator;
+  const addressStr = toHexString(withdrawalCredentials.slice(12));
```
Comment (twoeths) [inline packages/state-transition/src/block/processExecutionLayerWithdrawalRequest.ts]: use `subarray()` if possible, check other places too

File: packages/state-transition/src/block/processExecutionLayerWithdrawalRequest.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    const pendingPartialWithdrawal = ssz.electra.PartialWithdrawal.toViewDU({
+      index: validatorIndex,
+      amount: amountToWithdraw,
+      withdrawableEpoch,
+    });
+    state.pendingPartialWithdrawals.push(pendingPartialWithdrawal);
+  }
+}
+
+function isValidValidator(
```
Comment (twoeths) [inline packages/state-transition/src/block/processExecutionLayerWithdrawalRequest.ts]: this function name is confusing, maybe `isValidatorEligibleForWithdraw()` or something else?

File: packages/state-transition/src/block/processOperations.ts [hunk 6]
```diff
@@ -45,13 +52,35 @@ export function processOperations(
...
+    }
   }
 
   if (fork >= ForkSeq.capella) {
     for (const blsToExecutionChange of (body as capella.BeaconBlockBody).blsToExecutionChanges) {
       processBlsToExecutionChange(state as CachedBeaconStateCapella, blsToExecutionChange);
     }
   }
+
+  if (fork >= ForkSeq.electra) {
```
Comment (twoeths) [inline packages/state-transition/src/block/processOperations.ts:65]: this is not the same order to the spec, maybe just follow the same:
```python
    # [New in Electra:EIP7002:EIP7251]
    for_ops(body.execution_payload.withdrawal_requests, process_execution_layer_withdrawal_request)
    for_ops(body.execution_payload.deposit_receipts, process_deposit_receipt)  # [New in Electra:EIP6110]
    for_ops(body.consolidations, process_consolidation)  # [New in Electra:EIP7251]
```

File: packages/state-transition/src/block/processWithdrawals.ts [hunk 7]
```diff
@@ -1,19 +1,32 @@
...
+} from "../util/index.js";
 
 export function processWithdrawals(
-  state: CachedBeaconStateCapella,
+  fork: ForkSeq,
+  state: CachedBeaconStateCapella | CachedBeaconStateElectra,
   payload: capella.FullOrBlindedExecutionPayload
 ): void {
-  const {withdrawals: expectedWithdrawals} = getExpectedWithdrawals(state);
+  // const {withdrawals: expectedWithdrawals, partialWithdrawalsCount} = getExpectedWithdrawals(fork, state);
```
Comment (twoeths) [inline packages/state-transition/src/block/processWithdrawals.ts]: remove if not use

Comment (twoeths) [reply packages/state-transition/src/block/processWithdrawals.ts]: nvm, need to use it for below function when ssz is available
```python
state.pending_partial_withdrawals = state.pending_partial_withdrawals[partial_withdrawals_count:]
```

File: packages/state-transition/src/block/processWithdrawals.ts [hunk 8]
```diff
@@ -62,18 +80,52 @@ export function processWithdrawals(
...
+        validator.effectiveBalance >= MIN_ACTIVATION_BALANCE &&
+        balances.get(withdrawalIndex) > MIN_ACTIVATION_BALANCE
+      ) {
+        const balanceOverMinActivationBalance = BigInt(balances.get(withdrawalIndex) - MIN_ACTIVATION_BALANCE);
+        const withdrawableBalance =
+          balanceOverMinActivationBalance < withdrawal.amount ? balanceOverMinActivationBalance : withdrawal.amount;
+        withdrawals.push({
+          index: withdrawalIndex,
+          validatorIndex: withdrawal.index,
+          address: validator.withdrawalCredentials.slice(12),
```
Comment (twoeths) [inline packages/state-transition/src/block/processWithdrawals.ts]: use `subarray()` instead

File: packages/state-transition/src/util/electra.ts [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+export function isPartiallyWithdrawableValidator(
+  fork: ForkSeq,
+  validatorCredential: ValidatorInfo,
+  balance: number
+): boolean {
+  const {effectiveBalance, withdrawalCredentials: withdrawalCredential} = validatorCredential;
+
+  if (fork < ForkSeq.capella) {
+    throw new Error("Unsupported fork");
```
Comment (twoeths) [inline packages/state-transition/src/util/electra.ts]: add more details to this error message: fork, the function name

File: packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts [hunk 10]
```diff
@@ -555,8 +557,9 @@ function preparePayloadAttributes(
     suggestedFeeRecipient: feeRecipient,
   };
 
-  if (ForkSeq[fork] >= ForkSeq.capella) {
+  if (ForkSeq[fork] === ForkSeq.capella || ForkSeq[fork] === ForkSeq.electra) {
```
Comment (twoeths) [inline packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts]: merge issue? only need this `if (ForkSeq[fork] >= ForkSeq.capella) {`

File: packages/state-transition/src/block/processConsolidation.ts [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    );
+  }
+  if (!hasExecutionWithdrawalCredential(targetValidator.withdrawalCredentials)) {
+    throw new Error(
+      `Consolidation target validator does not have execution withdrawal credentials: targetIndex=${targetIndex}`
+    );
+  }
+
+  // Verify the same withdrawal address
+  const sourceWithdrawalAddress = toHexString(sourceValidator.withdrawalCredentials.slice(12));
```
Comment (twoeths) [inline packages/state-transition/src/block/processConsolidation.ts]: use `subarray()` for these 2 places

File: packages/state-transition/src/block/processDeposit.ts [hunk 12]
```diff
@@ -41,68 +49,132 @@ export function processDeposit(fork: ForkSeq, state: CachedBeaconStateAllForks,
...
+  }
+
+  if (fork < ForkSeq.electra) {
+    state.balances.push(amount);
+  } else if (fork >= ForkSeq.electra) {
+    state.balances.push(0);
+    const stateElectra = state as CachedBeaconStateElectra;
+    const pendingBalanceDeposit = ssz.electra.PendingBalanceDeposit.toViewDU({
+      index: validatorIndex,
+      amount: BigInt(amount),
```
Comment (twoeths) [inline packages/state-transition/src/block/processDeposit.ts:149]: is it needed to model `amount` as BigInt? in `processPendingBalanceDeposits()` we need to cast it to Number anyway:
```typescript
for (const deposit of state.pendingBalanceDeposits.getAllReadonly()) {
    const {amount} = deposit;
    if (processedAmount + amount > availableForProcessing) {
      break;
    }
    increaseBalance(state, deposit.index, Number(amount));
    processedAmount = processedAmount + amount;
    nextDepositIndex++;
  }
```

File: packages/state-transition/src/block/processDepositReceipt.ts [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+import {CachedBeaconStateElectra} from "../types.js";
+import {applyDeposit} from "./processDeposit.js";
+
+export function processDepositReceipt(
+  fork: ForkSeq,
+  state: CachedBeaconStateElectra,
+  depositReceipt: electra.DepositReceipt
+): void {
+  if (state.depositReceiptsStartIndex === UNSET_DEPOSIT_RECEIPTS_START_INDEX) {
+    state.depositReceiptsStartIndex = BigInt(depositReceipt.index);
```
Comment (twoeths) [inline packages/state-transition/src/block/processDepositReceipt.ts]: do not use BigInt unless it's really necessary due to performance reason

File: packages/state-transition/src/block/processDeposit.ts (hunk 3 above)
Comment (g11tech) [reply packages/state-transition/src/block/processDeposit.ts:83]: seems to be correct, so balance - MIN_ACTIVATION_BALANCE is the excess which would have been partially withdrawan, but is not queued to add to balance deposits

File: packages/state-transition/src/block/processDepositReceipt.ts (hunk 13 above)
Comment (g11tech) [reply packages/state-transition/src/block/processDepositReceipt.ts]: will review and do a followup of the the new amount types and indexes

**Review by g11tech — APPROVED** [review]
lgtm for the electra fork branch, any issues to be fixed will followup PRs on debugging or observations

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.22.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.22.0) :tada:

**Pull request merged**